# betaCateninQC.R

# Erich S. Huang
# Sage Bionetworks
# Seattle, Washington
# erich.huang@sagebase.org

#####
# LOAD IN REQUIRED LIBRARIES
#####

require(mGenomics)
require(snm)
require(ggplot2)
require(synapseClient)
mplotEnt <- loadEntity('syn274067')
attach(mplotEnt)
# synapseLogin('username', 'password')

#####
# LOAD IN DATA ENTITY
#####

bCatEnt <- loadEntity('syn138507')
fits <- runWorkflow(bCatEnt$cacheDir, workflow = 'snm')

#####
# PULL OUT THE EXPRESSION DATA
####

exprDat <- exprs(fits$hgu133a2[[1]])

# Create a treatment model matrix  (using the filename annotations)
treatment <- ifelse(grepl('bcat', list.files(bCatEnt$cacheDir)), "bCAT", "GFP")
X <- model.matrix(~ factor(treatment))
sigObj <- calcSig(exprDat, X)

# Generate a histogram
histFig <- qplot(sigObj$pval, geom = 'histogram') + 
  opts(title = 'P value distribution of transcripts') +
  xlab('p values') +
  ylab('Beta Catenin Effect on Gene Expression Variation')

# Take an SVD of the data. Look at eigenweights and first couple of eigengenes
svdObj <- fs(exprDat)
svdDF <- as.data.frame(cbind(1:19, svdObj$d, -1*svdObj$v))
colnames(svdDF) <- c('eigenGene', 'percentVariance', 
                     paste('prinComp', 1:19, sep = ''))
barFig <- ggplot(svdDF, aes(eigenGene, percentVariance)) +
  geom_bar(stat = 'identity') +
  opts(title = 'Percent Variance Explained by Each Eigengene') +
  xlab('Eigengene') +
  ylab('Percent Variance Explained')

eigenFig1 <- ggplot(svdDF, aes(eigenGene, prinComp1)) +
  geom_point(aes(colour = factor(treatment))) +
  opts(title = 'Eigengene 1 loadings by Sample') +
  xlab('Samples') +
  ylab('Eigengene 1 Loadings')

eigenFig2 <- ggplot(svdDF, aes(eigenGene, prinComp2)) +
  geom_point(aes(colour = factor(treatment))) +
  opts(title = 'Eigengene 2 loadings by Sample') +
  xlab('Samples') +
  ylab('Eigengene 2 Loadings')

#####
# REGRESS OUT THE KNOWN EXPERIMENTAL PERTURBATION EFFECT (TREATMENT)
#####

# In order to better understand perturbation-independent latent structure in
# the data

# Calculate the total sums of squares of the data.
datM <- rowMeans(exprDat)
datC <- sweep(exprDat, 1, datM)
tSSQ_dat <- sum(datC^2)

# Now we want to remove the effects of the biological treatment
# on the data.  The reason is we are interested in identifying 
# any latent structure. In order to do so we calcualte the
# residual sums of squares and take a singular value decomposition
# of the data.  Note the singular values are weighted to sum to 1.

residuals <- exprDat - t(X %*% solve(t(X) %*% X) %*% t(X) %*% t(exprDat))
rSSQ_dat <- sum(residuals^2)
u <- fs(residuals)
propSSQ <- round(rSSQ_dat * u$d,3) / tSSQ_dat
propSSQDF <- as.data.frame(cbind(1:19, propSSQ))
colnames(propSSQDF) <- c('eigenGene', 'propTSSQ')

propSSQFig <- ggplot(propSSQDF, aes(eigenGene, propSSQ)) +
  geom_bar(stat = 'identity') +
  opts(title = 'Proportion of total SSQ Explained by Each Eigengene') +
  xlab('Eigengene') +
  ylab('Proportion of total Sum of Squares Explained')

# Visual inspection of this figure suggests that the dependence kernel 
# should be set as rank 4

svaFit <- sva(exprDat, bio.var = X, n.sv = 4, num.iter = 30, diagnose = FALSE)
# Now, we'll take a look at the estimated basis vectors 

svaDF <- as.data.frame(cbind(1:19, svaFit$svd[[30]]$v))
colnames(svaDF) <- c('sample', paste('adjPrinComp', 1:4, sep = ''))

adjSVDFig1 <- ggplot(svaDF, aes(sample, adjPrinComp1)) +
  geom_point(aes(colour = factor(treatment))) +
  opts(title = 'Treatment Adjusted Eigengene 1 Loadings') +
  xlab('Sample') +
  ylab('Eigengene 1 Loading')

adjSVDFig2 <- ggplot(svaDF, aes(sample, adjPrinComp2)) +
  geom_point(aes(colour = factor(treatment))) +
  opts(title = 'Treatment Adjusted Eigengene 2 Loadings') +
  xlab('Sample') +
  ylab('Eigengene 1 Loading')

adjSVDFig3 <- ggplot(svaDF, aes(sample, adjPrinComp3)) +
  geom_point(aes(colour = factor(treatment))) +
  opts(title = 'Treatment Adjusted Eigengene 1 Loadings') +
  xlab('Sample') +
  ylab('Eigengene 3 Loading')

adjSVDFig4 <- ggplot(svaDF, aes(sample, adjPrinComp4)) +
  geom_point(aes(colour = factor(treatment))) +
  opts(title = 'Treatment Adjusted Eigengene 1 Loadings') +
  xlab('Sample') +
  ylab('Eigengene 4 Loading')
  
adjCompositeFig <- multiplot(adjSVDFig1,
                             adjSVDFig2,
                             adjSVDFig3,
                             adjSVDFig4, cols = 2)


# Plots 1 & 2 suggest that even with the experimental perturbation effect
# removed there is still latent structure that correlates with the 
# experimental treatment.

# One way to manages this is to estimate a dependence kernel from transcripts
# that are non-varying, as determined using the pi0 statistic to define a 
# significance cutoff. Basically, a dependence kernel is generated by taking
# an unweighted SVD of the data restricted to the first pi0 largest ranked 
# (by p values) transcripts.

# Get the null probes
nullProbes <- which(rank(1 - sigObj$pval) < (length(sigObj$pval) * sigObj$pi0))
u <- fs(exprDat[nullProbes,])
Z <- model.matrix(~ u$v[,1:4])
fits2 <- runWorkflow(bCatEnt$cacheDir,
                     workflow = "snm", bio.var = X, adj.var = Z, rm.adj = TRUE) 

dat2 <- exprs(fits2$hgu133a2[[1]])
u2 <- fs(dat2)

normDF <- as.data.frame(cbind(1:19, u2$v))
colnames(normDF) <- c('samples', paste('normPrinComp', 1:19, sep = ''))

normFig <- ggplot(normDF, aes(samples, normPrinComp1)) +
  geom_point(aes(colour = factor(treatment)))


par(mfrow=c(1,2))
hist(sig2$pval)
u2 <- fs(dat2)
barplot(u2$d, ylab="Prop Variance Explained by Each Eigengene")
plot(u2$v[,1], xlab="Samples",ylab="Eigengene 1")







